python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,112 @@
+import collections
+from typing import Dict, Iterator, List, Optional, Tuple
+
+from ray.types import ObjectRef
+from ray.data.block import Block, BlockMetadata
+from ray.data._internal.block_batching.interfaces import (
+    Batch,
+    BlockPrefetcher,
+)
+
+"""
+The algorithm uses both pipeline parallelism and data parallelism:
+
+If prefetch_batches=2, these are all the batches in flight:
+
+[User thread] trains on Batch 0
+   - [Fetch thread] Batch 1 in output queue
+        - [Worker thread 1] Batch 2 formatting + collating
+        - [Worker thread 2] Batch 3 formatting + collating
+        - [Raylet] Batches 4 + 5 fetched to local object store memory
+
+At any point in time there are prefetch_batches+1 batches in local heap memory.
+And the next set of prefetch_batches in local object store memory.
+
+The actual steps are as follows:
+
+In a single async thread, do the following:
+    1. Trigger Ray local prefetching of `prefetch_batches` worth of block object
+        references.
+    2. Resolve (i.e. call `ray.get()`) on the block references.
+    3. Perform the necessary batch slicing to construct full batches, possibly
+        shuffling if necessary.
+    4. Then, in a threadpool consisting of `prefetch_batches` threads:
+        3. Format the batches to the provided batch format.
+        4. Apply the collate function.
+    5. Fetch outputs from the threadpool, maintaining order of the batches.
+"""
+
+
+def prefetch_batches_locally(
+    block_ref_iter: Iterator[Tuple[ObjectRef[Block], BlockMetadata]],
+    prefetcher: BlockPrefetcher,
+    num_batches_to_prefetch: int,
+    batch_size: Optional[int],
+) -> Iterator[List[ObjectRef[Block]]]:
+    """Given an iterator of batched block references, returns an iterator over the same
+    block references while prefetching `num_batches_to_prefetch` batches in advance.
+
+    Args:
+        block_ref_iter: An iterator over batched block references.
+        prefetcher: The prefetcher to use.
+        num_batches_to_prefetch: The number of batches to prefetch ahead of the
+            current batch during the scan.
+        batch_size: User specified batch size, or None to let the system pick.
+    """
+
+    sliding_window = collections.deque()
+    current_window_size = 0
+
+    if batch_size:
+        num_rows_to_prefetch = num_batches_to_prefetch * batch_size
```

Comment:
Else set it to None?

Reply:
sure, we don't actually access it unless `batch_size` is not None